src/models/src/entity/atproto.rs

```diff
@@ -0,0 +1,368 @@
+        // the callback id for the cache should be inside the encrypted cookie
+        let callback_id = ApiCookie::from_req(req, COOKIE_UPSTREAM_CALLBACK).ok_or_else(|| {
+            ErrorResponse::new(
+                ErrorResponseType::Forbidden,
+                "Missing encrypted callback cookie",
+            )
+        })?;
+
+        // validate state
+        if callback_id != payload.state {
+            AuthProviderCallback::delete(callback_id).await?;
+
+            error!("`state` does not match");
+            return Err(ErrorResponse::new(
+                ErrorResponseType::BadRequest,
+                "`state` does not match",
+            ));
+        }
+        debug!("callback state is valid");
+
+        // validate csrf token
+        let slf = AuthProviderCallback::find(callback_id).await?;
+        if slf.xsrf_token != payload.xsrf_token {
+            AuthProviderCallback::delete(slf.callback_id).await?;
+
+            error!("invalid CSRF token");
+            return Err(ErrorResponse::new(
+                ErrorResponseType::Unauthorized,
+                "invalid CSRF token",
+            ));
+        }
+        debug!("callback csrf token is valid");
+
+        // request is valid -> fetch token for the user
+        let params = CallbackParams {
+            code: payload.code.clone(),
+            state: Some(payload.state.clone()),
+            iss: payload.iss.clone(),
+        };
+        // return early if we got any error
+        let (session_manager, _) = data.atproto.callback(params).await.map_err(|error| {
+            error!(%error, "failed to complete authorization callback for atproto");
+
+            ErrorResponse::new(
+                ErrorResponseType::BadRequest,
+                "failed to complete authorization callback for atproto",
+            )
+        })?;
+
+        let agent = Agent::new(session_manager);
+
+        let (did, email, email_confirmed) = match agent.api.com.atproto.server.get_session().await {
+            Ok(Object {
+                data:
+                    get_session::OutputData {
+                        did,
+                        email,
+                        email_confirmed,
+                        ..
+                    },
+                ..
+            }) => (did, email, email_confirmed),
+            Err(error) => {
+                error!(%error, "failed to get session for atproto");
+
+                return Err(ErrorResponse::new(
+                    ErrorResponseType::Internal,
+                    "failed to get session for atproto",
+                ));
+            }
+        };
+
+        let link_cookie = ApiCookie::from_req(req, PROVIDER_LINK_COOKIE)
+            .and_then(|value| AuthProviderLinkCookie::try_from(value.as_str()).ok());
+
+        if email.is_none() {
+            todo!()
+        }
+
+        let claims_user_id = did.to_string();
+
+        let user_opt = match User::find_by_federation("atproto", &claims_user_id).await {
```

Comment:
The hardcoded `provider_id` here could lead to issues in case someone created an `atproto` provider beforehand.

There 2 options:
- Define `atproto` as a blacklisted id for upstream providers in UI + API POST endpoint validation
- Create a section in the book for atproto setup and mention how an atproto provider needs to be set up and added manually

The latter would provide more control and freedom but could create issues of course (have not seen all code yet).

Reply:
I know it doesn't depend on auth providers, that's exactly the problem here.
Because if someone would add an upstream IdP with the id `atproto` manually, this would produce in conflicts.
You are using the `auth_provider_id` column here without it being a user-configured auth provider.

That's why I would simply blacklist `atproto` and mention it in the UI + API, that it's a reserved keyword, and all confusion would be gone. You can't misunderstand that.